### PR TITLE
Battle calc fixes and improvements

### PR DESCRIPTION
**Pull request by ron-murhammer**
## Overview
- Addresses BC improvements discussed here: https://forums.triplea-game.org/topic/369/battle-calculator-question-feature
- Fixes a few issues around the BC not handling isInfra and 0 move units properly

## Functional Changes
- Added ability to use CTRL+A and CTRL+D to add attackers and defenders to last active battle calc window
- Improved default player selection for attacker and defender when opening a new battle calc
- Allow changing defending player in BC and if any of the current units are owned by the new defending player then units no longer reset
- When switching players in BC, it only moves over the units of the new attacking player (not any allied units which can't actually attack)
- BC considers isInfra units in unit counts and TUV (otherwise you could add combat isInfra units and it would appear they would always improve your TUV chance even if they were pretty much always being lost)
- Can have 0 move and can't combat move units in BC attackers for maps which have limited combat rounds (this is because battles can last more than 1 round so these units can become attackers)
- Changed action tab hotkey to 'C' and updated help menu
- Defending units now create separate columns for each player
- Scales appropriately based on unit list and number of defending players but also has a max window size slightly smaller than TripleA window

## Manual Testing Performed
- Did testing on revised, global, TWW, dragon war, house of hapsburg, civil war

### Before
image

### After
image

## Additional Review Notes
- There was also a considerable amount of refactoring that I'll try to point out to avoid confusion with the actual changes

### CODE REVIEW

**Comment by ron-murhammer** [comment]
@RoiEXLab Thanks for your stream guidance as my stream kung fu isn't quite up to par yet and I still struggle with the double colon operator. I'll look to make those updates later today.

File: game-core/src/main/java/games/strategy/engine/data/UnitCollection.java [hunk 1]
```diff
@@ -167,6 +168,13 @@ public boolean isEmpty() {
     return count;
   }
 
+  public List<PlayerId> getPlayersByUnitCount() {
+    final IntegerMap<PlayerId> map = getPlayerUnitCounts();
+    final List<PlayerId> players = new ArrayList<>(map.keySet());
+    players.sort(Comparator.<PlayerId>comparingInt(p -> map.getInt(p)).reversed());
```
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply game-core/src/main/java/games/strategy/engine/data/UnitCollection.java]: Actually it does but I messed up the syntax 🙈

File: game-core/src/main/java/games/strategy/triplea/odds/calculator/AggregateResults.java [hunk 2]
```diff
@@ -66,7 +66,7 @@ public void addResults(final Collection<BattleResults> results) {
       return 0.0;
     }
     return results.stream()
-        .mapToDouble(BattleResults::getAttackingCombatUnitsLeft)
+        .mapToDouble(result -> result.getRemainingAttackingUnits().size())
```
[the remark this answers is not in the document]
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/odds/calculator/AggregateResults.java]: Updated. Though needs to be mapToDouble for division afterwards.

File: game-core/src/main/java/games/strategy/triplea/odds/calculator/AggregateResults.java [hunk 3]
```diff
@@ -134,7 +134,7 @@ public double getAverageTuvSwing(final PlayerId attacker, final Collection<Unit>
       return 0.0;
     }
     return results.stream()
-        .mapToDouble(BattleResults::getDefendingCombatUnitsLeft)
+        .mapToDouble(result -> result.getRemainingDefendingUnits().size())
```
[the remark this answers is not in the document]
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/odds/calculator/AggregateResults.java]: Updated. Though needs to be mapToDouble for division afterwards.

File: game-core/src/main/java/games/strategy/engine/data/UnitCollection.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply game-core/src/main/java/games/strategy/engine/data/UnitCollection.java]: I updated the original suggestion accordingly

[the remark this answers is not in the document]
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/engine/data/UnitCollection.java]: Ah ok. I see it now. Updated.

**Comment by ron-murhammer** [comment]
Alright. I think this is ready for review now. I'll probably add some screenshots but I think everything is working as intended and that I've had enough Swing punishment for a while.

**Pull request merged**